lib/service/service.go

```diff
@@ -1129,6 +1130,20 @@ func NewTeleport(cfg *servicecfg.Config) (*TeleportProcess, error) {
 			process.logger.WarnContext(process.ExitContext(), "Use of external upgraders on control-plane instances is not recommended.")
 		}
 
+		if upgraderKind == "unit" {
+			process.RegisterFunc("autoupdates.endpoint.export", func() error {
+				if err := endpoint.Export(process.ExitContext(), resolverAddr); err != nil {
```

Comment:
If we are joining via auth I think this will do weird things with the updater as it will set the endpoint override to `auth.example.com:3025` which is not a valid version server.

This is not a possible case for Cloud, but this can and will happen in self-hosted setups. I'm not sure we can catch all cases where we connect through a proxy, but known cases are:
- when `cfg.ProxyServer` is set
- (ugly) when port is 443 or 3080

We can also make this logic cloud-specific by looking for the `teleport.sh` suffix in the resolver address.

Reply:
I've made changes in [57cf444](https://github.com/gravitational/teleport/pull/42810/commits/57cf4448f6e576ace75eb338c83a49046c0e76ae) so that this export function will only be run for Cloud instances.

Additionally, the export function will attempt to verify the endpoint first before writing the config. So if the resolver address results in an invalid version server, the endpoint will not be written.